File: ruby/bob/bob.rb
class Bob

  def hey(message)
    analyzer = MessageAnalyzer.new(message)
    case
    when analyzer.silence?
      "Fine. Be that way!"
    when analyzer.shouting?
      "Woah, chill out!"
    when analyzer.question?
      "Sure."
    else
      "Whatever."
    end
  end

end

class MessageAnalyzer

  attr_reader :message

  def initialize(message)
    @message = message.to_s
  end

  def silence?
    message.empty?
  end

  def shouting?
    message == message.upcase
  end

  def question?
    message.end_with?("?")
  end

end

File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(input) do
    cond do
      silence?(input)  -> "Fine. Be that way!"
      shouting?(input) -> "Woah, chill out!"
      question?(input) -> "Sure."
      true             -> "Whatever."
    end
  end

  defp silence?(input) do
    String.strip(input) == ""
  end

  defp shouting?(input) do
    String.upcase(input) == input
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

end
